### ISSUE TITLE
[BUG] Machine pool info changes saved before UI complete during toggle from Form to YAML editor

### ISSUE THREAD

**Issue by gaktive** [rancher/rancher#43993]
**Rancher Server Setup**
- Rancher version: 2.8.0
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc):
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version:
- Cluster Type (Local/Downstream): 
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):
 * Infrastructure Provider = DigitalOcean & vSphere

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions:
  - Machine pool access

**Describe the bug**
UI has been doing work on fixing some YAML editor work and discovered in https://github.com/rancher/dashboard/issues/9697 that when editing a machine pool and then switching from Form mode to the YAML editor, changes are unexpectedly saved, even if the user cancels the operation. This behavior results in the machine pool values being updated when the user didn't intend to make any changes.

What looks to be happening per @momesgin is:
> The current process involves the Frontend triggering a Backend call with the updated configuration to obtain the updated YAML for rendering. This involves a PUT request that saves the cluster with the modified configuration.

This could be considered an enhancement but the proposed request comes from a bug: 

- Propose the creation of a new endpoint specifically designed to retrieve the updated YAML without immediately saving it.
- In this approach, the Frontend would use the new endpoint to fetch the updated YAML and render it. The actual saving (via PUT request) would only occur when the user explicitly clicks on the "Save" button.
- This approach eliminates the necessity for displaying a modal altogether, streamlining the user experience.

The UI could come up with its own solution based on detecting changes but the UX looks murky. The benefit of this proposal is that by requesting the machine config YAML that we can show inline when editing cluster YAML… and then save that YAML when the user clicks on saving the cluster makes it a better experience.

**To Reproduce**

1. Navigate to Cluster Management and click on "Create"
2. Provision a new RKE 2 Cluster (this has been reproduced with both DigitalOcean & VMware vSphere)
- Give the cluster a name and leave all other fields as default
- Click Create
- Wait for the cluster to enter a ready state
3. Edit the cluster that was provisioned in Step 2
- Change some values for a Machine Pool, for example update the Size of a Digital Ocean cluster from Basic: 4GB to Basic: 8GB
- Click on "Edit as YAML"
- Click Cancel

** Expected Result **

The Edit operation was canceled and all values are reverted to as they were before the editing began

** Actual Result **

Cluster enters an Updating state and Machine Pools values are updated even though Cancel was clicked

**Comment by Jono-SUSE-Rancher** [comment]
@snasovich - Let's discuss this one the next time we sync up. We should target for 2.9 to unblock UI.

**Comment by snasovich** [comment]
@gaktive / @momesgin , apologies for delayed response here.

TLDR here is I don't believe proposed approach for backend changes would work. And for that matter I can't think of another backend enhancement that won't require deep redesign of our data structures related to clusters - which we won't be able to accommodate anytime soon.
So, for that reason suggesting to close this issue as "Won't Do" and leave https://github.com/rancher/dashboard/issues/9697 to address the issue on UI side.

Let me elaborate more on this.

The crux of the issue is that "machine config" portion of machine pool details are stored in separate CRs (e.g. `rke-machine-config.cattle.io.digitaloceanconfigs`) and they are referenced by `provisioning.cattle.io.clusters` object, e.g.:
```yaml
apiVersion: provisioning.cattle.io/v1
kind: Cluster
...
spec:
...
  rkeConfig:
...
    machinePools:
      - controlPlaneRole: true
        machineConfigRef:
          kind: DigitaloceanConfig
          name: nc-sergey-rke2-ui-test-sergey-rke2-ui-test-all-shkrn
        name: sergey-rke2-ui-test-all
      - drainBeforeDelete: true
        machineConfigRef:
          kind: DigitaloceanConfig
          name: nc-sergey-rke2-ui-test-sergey-rke2-ui-test-worker-4xhrc
        name: sergey-rke2-ui-test-worker
...
```

So, if there are changes done solely to machine configs on machine pool(s), the `provisioning.cattle.io.clusters` object itself doesn't have any direct changes on it, so the proposed "new endpoint specifically designed to retrieve the updated YAML without immediately saving it" won't help as the "updated YAML" will be the same.

Note that currently, after user finishes updating cluster configuration on "Edit Cluster" form, the UI submits `PUT` requests to update all cluster's machine configs (e.g. to `v1/rke-machine-config.cattle.io.digitaloceanconfigs/fleet-default/(pool-name)`) and only then calls `PUT` on corresponding cluster (`v1/provisioning.cattle.io.clusters/fleet-default/(cluster-name)`).
So, for this use case where "Edit as YAML" is used on "Edit Cluster" form, I'm wondering if UI could somehow "queue up" any updates the user may have done to machine configs while on "Edit as YAML" form and then when "Save" is pressed submit then in the same sequence as currently done when clicking "Save" on "Edit Cluster" form? Please note that while on "Edit as YAML" form the user cannot make any changes to machine configs themselves so the queued up changes to machine configs will still be valid. Yes, they can remove existing machine pools and even theoretically add new ones but this is a separate issue that is not handled well either (just tried removing machine pool and ended up orphaning it upon saving the cluster).
The above will be purely UI-side change for which https://github.com/rancher/dashboard/issues/9697 it the tracking issue.

**Comment by momesgin** [comment]
@snasovich thank you for investigating the proposal for the backend. Based on your feedback we should look into the pure UI solutions.

**Comment by snasovich** [comment]
@momesgin , thank you. Based on this response, closing this as "Won't Fix". We can reopen if the direction changes, let me know.
